clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/security/LoginTest.java

```diff
@@ -0,0 +1,267 @@
+import static org.apache.kafka.common.config.internals.BrokerSecurityConfigs.SASL_MECHANISM_INTER_BROKER_PROTOCOL_CONFIG;
+import static org.apache.kafka.server.config.ServerConfigs.AUTHORIZER_CLASS_NAME_CONFIG;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class LoginTest {
+
+    private static final String USERNAME = "userA";
+    private static final String PASSWORD = "pwd";
+    private static final String CLIENT_ID = "test-login-client";
+    private static final String LISTENER_PREFIX = "listener.name.controller.";
+    private static final String EXTERNAL_PREFIX = "listener.name.external.";
+    private static final String MECHANISMS = "PLAIN";
+    private static final String MECHANISMS_PREFIX = "plain.";
+    private static final String SASL_JAAS = "org.apache.kafka.common.security.plain.PlainLoginModule required "
+        + "user_" + USERNAME + "=\"" + PASSWORD + "\";";
+
+    @ClusterTest(
+        types = {Type.CO_KRAFT},
+        controllerSecurityProtocol = SecurityProtocol.SASL_PLAINTEXT,
+        brokerSecurityProtocol = SecurityProtocol.SASL_PLAINTEXT,
+        serverProperties = {
+            @ClusterConfigProperty(key = LISTENER_PREFIX + MECHANISMS_PREFIX + SASL_LOGIN_CLASS, value = "org.apache.kafka.clients.security.LoginTest$CustomerLogin"),
+            @ClusterConfigProperty(key = EXTERNAL_PREFIX + MECHANISMS_PREFIX + SASL_LOGIN_CLASS, value = "org.apache.kafka.clients.security.LoginTest$CustomerLogin"),
+            @ClusterConfigProperty(key = SASL_ENABLED_MECHANISMS_CONFIG, value = MECHANISMS),
+            @ClusterConfigProperty(key = SASL_MECHANISM_INTER_BROKER_PROTOCOL_CONFIG, value = MECHANISMS),
+            @ClusterConfigProperty(key = LISTENER_PREFIX + MECHANISMS_PREFIX + SASL_JAAS_CONFIG, value = SASL_JAAS),
+            @ClusterConfigProperty(key = AUTHORIZER_CLASS_NAME_CONFIG, value = "org.apache.kafka.clients.security.LoginTest$CustomerAuthorizer"),
+        }
+    )
+    public void testCustomLoginWithKafkaCluster(ClusterInstance cluster) {
+        try (Admin admin = cluster.admin(saslConfig())) {
+            int found = 0;
+            for (Metric metric : admin.metrics().values()) {
+                found += assertMetricName(
+                    metric.metricName(),
+                    expectedTags(Map.of("client-id", CLIENT_ID))
+                );
+            }
+            assertEquals(1, found, "Expected to find 1 metric");
+
+            Map<MetricName, KafkaMetric> allMetrics = Stream.of(
+                cluster.controllers().get(0).metrics().metrics(),
+                cluster.brokers().get(0).metrics().metrics()
+            ).collect(HashMap::new, Map::putAll, Map::putAll);
+            assertMetrics(
+                allMetrics,
+                expectedTags(Map.of(
+                    "mechanism", MECHANISMS,
+                    "listener", "EXTERNAL",
+                    "networkProcessor", "0"
```

Comment:
Should we check for all network processors instead of just 0?

Reply:
I’m not sure I fully understand why `num.io.threads` can influence the test. I traced the code and found that processor 0 acts as the acceptor in SocketServer. Could you share more thoughts on this?

https://github.com/apache/kafka/blob/cc4986e485e963c7fbac0bea7ca0e9b1b904e285/core/src/main/scala/kafka/server/BrokerServer.scala#L269

Reply:
Looking at the code, LoginManager, and hence Login, are shared across all processors for the same listener. You can see in `LoginManager.acquireLoginManager()` that it returns an existing instance if there's one already. This explains why we always see networkProcessor set to 0. So it makes no sense to have the processor Id in the tags.